scanpy/tools/_score_genes.py

```diff
@@ -155,12 +155,24 @@ def score_genes(
     control_genes = pd.Index([], dtype="string")
 
     # now pick `ctrl_size` genes from every cut
-    for cut in np.unique(obs_cut.loc[gene_list]):
-        r_genes: pd.Index[str] = obs_cut[obs_cut == cut].index
+    for cut in obs_cut.loc[gene_list].unique():
+        r_genes: pd.Index[str] = obs_cut[
+            (obs_cut == cut) & ~obs_cut.index.isin(gene_list)
+        ].index
+        if len(r_genes) == 0:
+            logg.warning(
+                f"No control genes for this cut. You might want to increase "
+                f"ctrl_size={ctrl_size} to sample from more control genes."
+            )
```

Comment:
What do you mean here? `ctrl_size` only comes into play in the next line, so if the cut is empty, it can’t be fixed by increasing `ctrl_size`, right?

Reply:
The condition for the warning being shown has nothing to do with `ctrl_size`, yet the warning talks about `ctrl_size`. Therefore following the recommendation in the warning (increasing `ctrl_size`) won’t ever be able to make the warning go away, which is confusing.

Ideally, the warning would only recommend doing a thing if doing the thing has the potential to make the warning go away. Is that possible?

Reply:
ah, I see what you mean 😅 You're right, I think the warning isn't necessary actually. The cut can only be empty if the user provides too many genes for scoring and there aren't many control genes left. Do you think it's worth putting a warning for that case or would you prefer no warning at all?

Reply:
If we can detect a problem and show a warning that actually has an actionable suggestion, that’s always a good idea.